### DISCUSSION TITLE
insert mode `Ctrl+left/right` navigation is kinda broken

### DISCUSSION THREAD

**Discussion by adsick** [helix-editor/helix#10266] [General]
Hi, referring [my long complain](https://github.com/helix-editor/helix/discussions/3388) about Helix not being as good as I would want it to be I am writing this issue to clear more on what is wrong with word-wise navigation in insert mode.

assume we have this text (repeated 2-3 times to demonstrate wrapping behaviors):
```
    println!("Hello, world!");
```

How it works in VS Code
---
```
    |println!("Hello, world!");
    println|!("Hello, world!");
    println!("|Hello, world!");
    println!("Hello|, world!");
    println!("Hello,| world!");
    println!("Hello, world|!");
    println!("Hello, world!");|
```
and backwards:
```
    println!("Hello, world!");|
    println!("Hello, world|!");
    println!("Hello, |world!");
    println!("|Hello, world!");
    println|!("Hello, world!");
    |println!("Hello, world!");
```

How it looks in Helix
---
```
    |println!("Hello, world!");
    printl|n!("Hello, world!");
    println!(|"Hello, world!");
    println!("Hell|o, world!");
    println!("Hello,| world!");
    println!("Hello, worl|d!");
    println!("Hello, world!")|;
```
you might think that the next step will be
```
    println!("Hello, world!");|
```
but in fact it wraps to the next line and sits like this:
```
    println!("Hello, world!");
    printl|n!("Hello, world!");
```
and when navigating backwards it will visit different stops than when navigating forward:
```
    println!("Hello, world!");
    printl|n!("Hello, world!");
```
```
    println!("Hello, world!");
    |println!("Hello, world!");
```

```
    println!("Hello, world|!");
    println!("Hello,| world!");
    println!("Hello|, world!");
    println!("|Hello, world!");
    println|!("Hello, world!");
    |println!("Hello, world!");
```
whoa, that's a lot of greetings! I really hope that you can spot what I'm talking about.

### from this little experiment we can conclude a few things:
1. Helix's word-wise navigation does not matches VSC's one (and also vim's one but you can check that for yourself)
there are other contexts like command-line prompt when you can use Ctrl+left/right to navigate, and they're also differ, but some things (like where the stop sits, on the last char or before it) are more sensible there as well
2. I find VS Code's variant much more sensible and convenient
3. Helix's navigation looks better backwards than forwards
4. Some deeper investigation needed to determine what to do with it.

I've noticed that Ctrl+left/right uses `move_prev_word_end` and `move_next_word_start` actions, and as properly mentioned in a similar issue #2729 it should be decided what we should `count as a word` (and whether to use or not this ambiguous term at all)

**Comment by archseer** [comment]
Probably has the same problem/fix as https://github.com/helix-editor/helix/pull/4392

**Comment by adsick** [comment]
> Probably has the same problem/fix as #4392

Ctrl+left/right is not working at master (nothing happens when I press it).

**Comment by adsick** [comment]
I can explain what's wrong here in fewer words:

let's say you have text "Helix is a nice editor."
if we're navigating it in VS Code using Ctrl+right the cursor always sits right after the end of the word like so:
```
Helix| is a nice editor.
Helix is| a nice editor.
Helix is a| nice editor.
Helix is a nice| editor.
Helix is a nice editor|.
```
(notice the last one)
but in Helix (using w in normal mode) it will be
```
...the same, but then...
Helix is a nice| editor.
Helix is a nice edito|r.
```
which is kinda sus

**Comment by archseer** [comment]
I'm saying it can probably be fixed in the same way as #4392, that PR only touched `delete_word_backward`.

**Comment by peterjc** [comment]
How does this work for you, based on how #5009 proposed remapping ctrl+w/e/b:

```
C-left = ["move_prev_word_start" , "collapse_selection"]
C-right = ["move_next_word_end", "move_char_right"]
```

For me this then matches ctrl+left/right in Sublime Text, alt+left/right in Sublime Text, alt+left/right in Apple's TextEdit, etc. It feels very familiar and consistent for me!

(Tested with Helix 25.1 on macOS)
